1.Pivot Tables	
    Motivating Pivot Tables
	Pivot Tables by Hand
	Pivot Table Syntax



In [11]:
import numpy as np
import pandas as pd
import seaborn as sb
titanic=sb.load_dataset('titanic')
#print(titanic)
print(titanic.groupby('sex')[['survived']].mean())
print(titanic.groupby(['sex','class'])['survived'].mean()) 

#Pivot_table keyword
print(titanic.pivot_table('survived',index='sex',columns='class'))
print(titanic.pivot_table(index='sex',columns='class',aggfunc={'survived':sum,'fare':'mean'}))

        survived
sex             
female  0.742038
male    0.188908
sex     class 
female  First     0.968085
        Second    0.921053
        Third     0.500000
male    First     0.368852
        Second    0.157407
        Third     0.135447
Name: survived, dtype: float64
class      First    Second     Third
sex                                 
female  0.968085  0.921053  0.500000
male    0.368852  0.157407  0.135447
              fare                       survived             
class        First     Second      Third    First Second Third
sex                                                           
female  106.125798  21.970121  16.118810       91     70    72
male     67.226127  19.741782  12.661633       45     17    47


2.Vectorized String Operations	
    Introducing Pandas String Operations
	Tables of Pandas String Methods
    
    
len()	lower()	translate()	islower()
ljust()	upper()	startswith()	isupper()
rjust()	find()	endswith()	isnumeric()
center()	rfind()	isalnum()	isdecimal()
zfill()	index()	isalpha()	split()
strip()	rindex()	isdigit()	rsplit()
rstrip()	capitalize()	isspace()	partition()
lstrip()	swapcase()	istitle()	rpartition()

In [21]:
import pandas as pd
a=['datascience','java','python','django']
[i.capitalize() for i in a] #=>Common way of capitalizing
ser=pd.Series(a)
print(ser)
print(ser.str.capitalize()) #=>By pandas
print(ser.str.lower())
print(ser.str.len())
print(ser.str.startswith('p'))
zz='testing###'
print(zz.zfill(10))
print(zz.ljust(10,'@'))
print(zz.isalpha())

0    datascience
1           java
2         python
3         django
dtype: object
0    Datascience
1           Java
2         Python
3         Django
dtype: object
0    datascience
1           java
2         python
3         django
dtype: object
0    11
1     4
2     6
3     6
dtype: int64
0    False
1    False
2     True
3    False
dtype: bool
testing###
testing###
False


In [27]:
a='               tTSt                 '
print(a)
print(a.strip())
print(a.rstrip())
print(a.lstrip())
print(a.swapcase())
t='Data Size'
print(t.istitle())

               tTSt                 
tTSt
               tTSt
tTSt                 
               TtsT                 
True


3.Working with Time Series	
    Working with Time Series
    1.Time stamps (eg: january 31st,2023 at 8.25PM)
    2.Time interval (eg: 24 hours in a day)
    3.Time deltas (eg: 22.56 seconds)
	Dates and Times in Python
    
	Pandas Time Series: Indexing by Time
	Pandas Time Series Data Structures
	Frequencies and Offsets
	Resampling, Shifting, and Windowing
    
    
    
Code	Meaning	Time span (relative)	Time span (absolute)
Y	Year	± 9.2e18 years	[9.2e18 BC, 9.2e18 AD]
M	Month	± 7.6e17 years	[7.6e17 BC, 7.6e17 AD]
W	Week	± 1.7e17 years	[1.7e17 BC, 1.7e17 AD]
D	Day	± 2.5e16 years	[2.5e16 BC, 2.5e16 AD]
h	Hour	± 1.0e15 years	[1.0e15 BC, 1.0e15 AD]
m	Minute	± 1.7e13 years	[1.7e13 BC, 1.7e13 AD]
s	Second	± 2.9e12 years	[ 2.9e9 BC, 2.9e9 AD]
ms	Millisecond	± 2.9e9 years	[ 2.9e6 BC, 2.9e6 AD]
us	Microsecond	± 2.9e6 years	[290301 BC, 294241 AD]
ns	Nanosecond	± 292 years	[ 1678 AD, 2262 AD]
ps	Picosecond	± 106 days	[ 1969 AD, 1970 AD]
fs	Femtosecond	± 2.6 hours	[ 1969 AD, 1970 AD]
as	Attosecond	± 9.2 seconds	[ 1969 AD, 1970 AD]

In [30]:
import numpy as np
date = np.array('2015-07-04', dtype=np.datetime64)
date

array('2015-07-04', dtype='datetime64[D]')

In [4]:
#Dates and Times in Python
# datetime & dateutil
from datetime import datetime
print(datetime(year=2023, month=1, day=31))

from dateutil import parser
date=parser.parse("31st of January, 2023")
print(date)

#finding the day 
date.strftime('%A')

2023-01-31 00:00:00
2023-01-31 00:00:00


'Tuesday'

In [12]:
#Pandas Time Series: Indexing by Time

import numpy as np
date=np.array('2015-07-04', dtype=np.datetime64)
print(date)
date2=np.datetime64('2023-01-31 12:59:59.50')
print(date2)

import pandas as pd
d=pd.to_datetime('31st of January, 2023')
print(d)
print(d.strftime('%A'))

#DatetimeIndex
ind=pd.DatetimeIndex(['2014-07-04','2014-08-04','2015-07-04','2015-08-04'])
i1=pd.Series([0,1,2,3],index=ind)
print(i1)

print(i1['2014-07-04':'2015-07-04']) # Printing the data from and to 
print(i1['2014']) #Print the date by year

2015-07-04
2023-01-31T12:59:59.500
2023-01-31 00:00:00
Tuesday
2014-07-04    0
2014-08-04    1
2015-07-04    2
2015-08-04    3
dtype: int64
2014-07-04    0
2014-08-04    1
2015-07-04    2
dtype: int64
2014-07-04    0
2014-08-04    1
dtype: int64


#Pandas Time Series Data Structures
    1.Time stamps (eg: january 31st,2023 at 8.25PM)=>numpy.datetime64, DatetimeIndex
    2.Time interval (eg: 24 hours in a day)=>numpy.datetime64,PeriodIndex
    3.Time deltas (eg: 22.56 seconds)=>numpy.timedelta64, TimedeltaIndex
    

In [16]:
#Time stamp

import pandas as pd
d=pd.to_datetime('31st of January, 2023')
print(d)

ind=pd.DatetimeIndex(['2014-07-04','2014-08-04','2015-07-04','2015-08-04'])
i1=pd.Series([0,1,2,3],index=ind)
print(i1)

# Time Interval 

print(d.to_period('D'))

# Time Deltas 

print(pd.timedelta_range(0, periods=10, freq='H'))


2023-01-31 00:00:00
2014-07-04    0
2014-08-04    1
2015-07-04    2
2015-08-04    3
dtype: int64
2023-01-31
TimedeltaIndex(['0 days 00:00:00', '0 days 01:00:00', '0 days 02:00:00',
                '0 days 03:00:00', '0 days 04:00:00', '0 days 05:00:00',
                '0 days 06:00:00', '0 days 07:00:00', '0 days 08:00:00',
                '0 days 09:00:00'],
               dtype='timedelta64[ns]', freq='H')


Frequencies and Offsets

Code	Description	Code	Description
D	Calendar day	B	Business day
W	Weekly		
M	Month end	BM	Business month end
Q	Quarter end	BQ	Business quarter end
A	Year end	BA	Business year end
H	Hours	BH	Business hours
T	Minutes		
S	Seconds		
L	Milliseonds		
U	Microseconds		
N	nanoseconds	

In [17]:
dti = pd.date_range("2018-01-01", periods=3, freq="H")
print(dti)

DatetimeIndex(['2018-01-01 00:00:00', '2018-01-01 01:00:00',
               '2018-01-01 02:00:00'],
              dtype='datetime64[ns]', freq='H')


In [22]:
d=pd.Timestamp('2023-01-31')
print(d.day_name())
d1=d+pd.Timedelta("1 day")
print(d1.day_name())

print(pd.offsets.BDay())
a=d+pd.offsets.BDay()
print(a)
print(a.day_name())

Tuesday
Wednesday
<BusinessDay>
2023-02-01 00:00:00
Wednesday


In [24]:
#Resampling, Shifting, and Windowing

In [28]:
pip install pandas-datareader

Note: you may need to restart the kernel to use updated packages.
